File: src/main/java/moe/knox/factorio/intellij/util/FilesystemUtil.java
package moe.knox.factorio.intellij.util;

import com.intellij.ide.plugins.cl.PluginAwareClassLoader;
import com.intellij.openapi.application.PathManager;
import org.jetbrains.annotations.NotNull;

import java.nio.file.Path;

public class FilesystemUtil {
    public static @NotNull Path getPluginDir() {
        return PathManager.getPluginsDir().resolve(getPluginId());
    }

    public static String getPluginId() {
        return ((PluginAwareClassLoader) FilesystemUtil.class.getClassLoader()).getPluginDescriptor().getPluginId().getIdString();
    }
}

File: src/main/kotlin/io/serieznyi/intellij/factorioapicompletion/intellij/FilesystemUtil.kt
package io.serieznyi.intellij.factorioapicompletion.intellij

import com.intellij.ide.plugins.PluginManagerCore
import com.intellij.openapi.extensions.PluginId
import io.serieznyi.intellij.factorioapicompletion.core.util.io.findOrCreateDirectory
import java.nio.file.Path

class FilesystemUtil {
    companion object {
        private val PLUGIN_ID: PluginId = PluginId.getId("io.serieznyi.factorio-api-completion")

        fun pluginDir(): Path {
            val descriptor = PluginManagerCore.getPlugin(PLUGIN_ID) ?: throw RuntimeException("Unexpected error. Plugin dir not found")

            return descriptor.pluginPath
        }

        fun cacheDir(): Path {
            val pluginDir = this.pluginDir()

            return pluginDir.resolve("cache").findOrCreateDirectory()
        }

        fun apiVersionsDir(): Path {
            val pluginDir = this.pluginDir()

            return pluginDir.resolve("cache").findOrCreateDirectory()
        }
    }
}